modules/uhttp/filters.go

```diff
@@ -48,76 +48,83 @@ func (f FilterFunc) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request
-		next.ServeHTTP(fxctx, w, r)
-	}
+type tracingServerFilter struct {
+	scope tally.Scope
 }
 
-func tracingServerFilter(host service.Host) FilterFunc {
-	return func(ctx context.Context, w http.ResponseWriter, r *http.Request, next Handler) {
-		// TODO:(anup) GFM-257 benchmark performance comparing with using type assertion
-		fxctx := &fxcontext.Context{
-			Context: ctx,
-		}
-		operationName := r.Method
-		carrier := opentracing.HTTPHeadersCarrier(r.Header)
-		spanCtx, err := opentracing.GlobalTracer().Extract(opentracing.HTTPHeaders, carrier)
-		if err != nil && err != opentracing.ErrSpanContextNotFound {
-			fxctx.Logger().Info("Malformed inbound tracing context: ", "error", err.Error())
-		}
-		span := opentracing.GlobalTracer().StartSpan(operationName, ext.RPCServerOption(spanCtx))
-		ext.HTTPUrl.Set(span, r.URL.String())
-		defer func() {
-			span.Finish()
-			fxctx.Logger().Debug("Span finished")
-		}()
-
-		fxctx = fxctx.WithContextAwareLogger(span)
-		fxctx = &fxcontext.Context{
-			Context: opentracing.ContextWithSpan(fxctx, span),
-		}
-		r = r.WithContext(fxctx)
-		next.ServeHTTP(fxctx, w, r)
+func (f tracingServerFilter) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request, next Handler) {
+	stopWatch := f.scope.Timer("tracing").Start()
+	operationName := r.Method
+	carrier := opentracing.HTTPHeadersCarrier(r.Header)
+	spanCtx, err := opentracing.GlobalTracer().Extract(opentracing.HTTPHeaders, carrier)
+	if err != nil && err != opentracing.ErrSpanContextNotFound {
+		ctx.Logger().Info("Malformed inbound tracing context: ", "error", err.Error())
 	}
+	span := opentracing.GlobalTracer().StartSpan(operationName, ext.RPCServerOption(spanCtx))
+	ext.HTTPUrl.Set(span, r.URL.String())
+	defer func() {
+		span.Finish()
+		ctx.Logger().Debug("Span finished")
+	}()
+
+	ctx = ctx.(*fxcontext.Context).WithContextAwareLogger(span)
+	ctx = &fxcontext.Context{
+		Context: opentracing.ContextWithSpan(ctx, span),
+	}
+	r = r.WithContext(ctx)
+	stopWatch.Stop()
+	next.ServeHTTP(ctx, w, r)
 }
 
 // authorizationFilter authorizes services based on configuration
-func authorizationFilter(host service.Host) FilterFunc {
-	return func(ctx context.Context, w http.ResponseWriter, r *http.Request, next Handler) {
-		fxctx := &fxcontext.Context{
-			Context: ctx,
-		}
+type authorizationFilter struct {
+	authClient auth.Client
+	scope      tally.Scope
+}
 
-		if err := host.AuthClient().Authorize(fxctx); err != nil {
-			host.Metrics().SubScope("http").SubScope("auth").Counter("fail").Inc(1)
-			fxctx.Logger().Error(auth.ErrAuthorization, "error", err)
-			w.WriteHeader(http.StatusUnauthorized)
-			fmt.Fprintf(w, "Unauthorized access: %+v", err)
-			return
-		}
-		next.ServeHTTP(fxctx, w, r)
+func (f authorizationFilter) Apply(ctx fx.Context, w http.ResponseWriter, r *http.Request, next Handler) {
+	stopWatch := f.scope.Timer("auth").Start()
+	if err := f.authClient.Authorize(ctx); err != nil {
+		f.scope.SubScope("auth").Counter("fail").Inc(1)
```

Comment:
subscopes should never be created in the hot loop - it's a pretty expensive operation.

You can see here that it does a lot of copying: https://github.com/uber-go/tally/blob/master/scope.go#L309

We should either cache the sub-scope on some object, or figure out a way to emit metrics without creating subscopes for each filter operation